mirrord/config/src/feature/network/filter.rs

```diff
@@ -0,0 +1,415 @@
+                        .or_else(|_| format!("[{address}]:{port}").parse())
+                        .map(AddressFilter::Socket)
+                        // Neither IPv4 nor IPv6, it's probably a name.
+                        .unwrap_or(AddressFilter::Name(address.to_string(), port))
+                },
+                |subnet| AddressFilter::Subnet(subnet, port),
+            );
+
+        if rest.is_empty() {
+            Ok(filter)
+        } else {
+            Err(Self::Err::TrailingValue(rest.to_string()))
+        }
+    }
+}
+
+/// <!--${internal}-->
+/// The parsed filter with its [`ProtocolFilter`] and [`AddressFilter`].
+#[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Ord, Hash)]
+pub struct ProtocolAndAddressFilter {
+    /// Valid protocol types.
+    pub protocol: ProtocolFilter,
+
+    /// Address|name|subnet we're going to filter.
+    pub address: AddressFilter,
+}
+
+#[derive(Error, Debug)]
+pub enum ProtocolAndAddressFilterError {
+    #[error(transparent)]
+    Address(#[from] AddressFilterError),
+    #[error(transparent)]
+    Protocol(#[from] ParseProtocolError),
+}
+
+impl From<nom::Err<nom::error::Error<&str>>> for ProtocolAndAddressFilterError {
+    fn from(value: nom::Err<nom::error::Error<&str>>) -> Self {
+        Self::Address(value.into())
+    }
+}
+
+impl FromStr for ProtocolAndAddressFilter {
+    type Err = ProtocolAndAddressFilterError;
+
+    fn from_str(input: &str) -> Result<Self, Self::Err> {
+        // Perform the basic parsing.
+        let (rest, protocol) = protocol(input)?;
+        let protocol = protocol.parse()?;
+
+        let address = rest.parse()?;
+
+        Ok(Self { protocol, address })
+    }
+}
+
+/// <!--${internal}-->
+///
+/// Parses `tcp://`, extracting the `tcp` part, and discarding the `://`.
+fn protocol(input: &str) -> IResult<&str, &str> {
+    let (rest, protocol) = opt(terminated(take_until("://"), tag("://")))(input)?;
+    let protocol = protocol.unwrap_or("any");
+
+    Ok((rest, protocol))
+}
+
+/// <!--${internal}-->
+///
+/// We try to parse 3 different kinds of values here:
+///
+/// 1. `name.with.dots`;
+/// 2. `1.2.3.4.5.6`;
+/// 3. `[dad:1337:fa57::0]`
+///
+/// Where 1 and 2 are handled by `dotted_address`.
+///
+/// The parser is not interested in only eating correct values here for hostnames, ip addresses,
+/// etc., it just tries to get a good enough string that could be parsed by
+/// `SocketAddr::parse`, or `IpNet::parse`.
+///
+/// Returns `0.0.0.0` if it doesn't parse anything.
+fn address(input: &str) -> IResult<&str, String> {
+    let ipv6 = many1(alt((alphanumeric1, tag(":"))));
+    let ipv6_host = delimited(tag("["), ipv6, tag("]"));
+
+    let host_char = alt((alphanumeric1, tag("-"), tag("_"), tag(".")));
+    let dotted_address = many1(host_char);
+
+    let (rest, address) = opt(alt((dotted_address, ipv6_host)))(input)?;
+
+    let address = address
+        .map(|addr| addr.concat())
+        .unwrap_or(String::from("0.0.0.0"));
```

Comment:
Is it worth adding a trace log here if the parser failed to parse and returned 0.0.0.0?

Reply:
This actually made me visit the parsing logic, which previously I just copied. It makes no sense to provide a default here, it's flaky. Changed `AddressFilter::from_str` a bit to be more explicit and catch dumb cases